### PR TITLE
Make CORS domains configurable in UI

### LINKED ISSUE #13156: Let super users configure CORS hostnames in the UI
Currently CORS headers can be configured in the config file, as documented in our FAQ: 
https://matomo.org/faq/how-to/faq_18694/

Because configuring these hostnames is a quite frequent operation, we would like to offer this in the user interface in the settings. There is a similar feature and UI control used for defining the trusted hostsnames in the settings.

This setting would be available only for Super users.

### PR DESCRIPTION
**Pull request by sgiehl**
The new tuple input fields kind of did the same as the configuration for trusted host.
But to be able to provide the current value in the format the JS needs it, I added a new callback to the field config to convert the value. It's currently called `prepareValue`, but I'm not very happy with that name, but couldn't think of a better one. Maybe someone has an idea...

Note: By moving the trusted host setting to the SystemSetting class there is a change in UI. Before the was a confirmation box shown before changing the trusted hosts. Imho it's not really needed, but we could also implement an optional confirmation for all plugin settings (that can be activated in each SystemSettings class)

fixes #13156

### CODE REVIEW

File: core/Settings/Setting.php [hunk 1]
```diff
@@ -201,7 +201,17 @@ public function save()
      */
     public function getValue()
     {
-        return $this->storage->getValue($this->name, $this->defaultValue, $this->type);
+        $value = $this->storage->getValue($this->name, $this->defaultValue, $this->type);
+
+        $config = $this->configureField();
```
Comment (tsteur) [inline core/Settings/Setting.php]: A while ago we refactored the entire settings API so we basically wouldn't need to `configureField()` when getting the value as it is a performance issue. Neither should we `validateValue` ideally. Any chance we can tweak the UI component to return the value in an alternative format through a flag or something? 
We'd really need to make sure to leave the `getValue` as it is... I probably should have added a comment about this...

File: plugins/CoreAdminHome/SystemSettings.php [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            $field1 = new FieldConfig\MultiPair(Piwik::translate('Overlay_Domain'), 'domain', FieldConfig::UI_CONTROL_TEXT);
+            $field->uiControlAttributes['field1'] = $field1->toArray();
+            $field->transform = function($values) {
+                $corsDomains = [];
+                foreach ($values as $value) {
+                    if (!empty($value['domain'])) {
+                        $corsDomains[] = $value['domain'];
+                    }
+                }
+                return $corsDomains;
```
Comment (diosmosis) [inline plugins/CoreAdminHome/SystemSettings.php]: I think this could be shortened to `return array_column($values, 'domain');`

Comment (diosmosis) [reply plugins/CoreAdminHome/SystemSettings.php]: Same for `$hostnames` below

File: core/Settings/FieldConfig.php [hunk 3]
```diff
@@ -214,6 +214,19 @@ class FieldConfig
...
+     *
+     * **Example**
+     *
+     *     $setting->prepareValue = function ($value, Setting $setting) {
+     *         return split(',', $value);
+     *     }
+     *
+     * @var null|\Closure
+     */
+    public $prepareValue = null;
```
Comment (diosmosis) [inline core/Settings/FieldConfig.php]: Not sure if it's better or worse (you can be the judge), but what about naming this variable `$decodeValue`?

**Comment by tsteur** [comment]
> multi tuple but with only one field and thus returns the data as a simple array

sounds good to me 👍 be also reusable then etc.

**Comment by sgiehl** [comment]
> Here I would maybe only write "hostname" instead of "valid matomo hostname"?

That was this way before. Could change it, but that needs new translations then...

Everthing else should be updated now

**Comment by sgiehl** [comment]
@tsteur I've reorganized the form demos a bit. Let me know what you think...

**Comment by tsteur** [comment]
Looks great 👍  for some reason I couldn't see the failing UI test in https://builds-artifacts.matomo.org/matomo-org/matomo/corssettings/28884/ ... the tests would still need to be fixed I reckon (unless I missed it in the changes files in the PR)

**Comment by sgiehl** [comment]
tests should now be updated. Not sure why the UI tests didn't fail before...

**Comment by sgiehl** [comment]
So guess we are good to merge? Could anybody confirm that it's fine the trusted hosts confirmation will be removed?

**Comment by tsteur** [comment]
For me it would be fine to merge and that the confirmation is removed. Not really needed IMO

**Comment by diosmosis** [comment]
Noticed something odd in the config file diagnostic page: https://builds-artifacts.matomo.org/matomo-org/matomo/corssettings/28915/UIIntegrationTest_admin_diagnostics_configfile.png

The cors_domain/trusted_hosts settings are now in the `CoreAdminHome` section as well as `General`. Is that expected?

**Comment by sgiehl** [comment]
> If possible it's slightly better to keep it, but it's also not a blocker/required. How much effort would it be to keep it?

we would need to extend the default behavior for system settings provided by plugins to implement something like that

**Comment by sgiehl** [comment]
> The cors_domain/trusted_hosts settings are now in the CoreAdminHome section as well as General. Is that expected?

Need to have a look at that. Imho it doesn't make sense to list system settings there that are of type `ManagedInConfigOnly`

**Comment by diosmosis** [comment]
👍 looks good, will merge

**Comment by diosmosis** [comment]
@sgiehl Actually spoke too soon, there is another failing test in ConfigReaderTest. Tried to fix myself but wasn't able to say whether the changes to `addConfigValuesFromSystemSettings()` were correct or not. Maybe the `!empty($configValues[$configSection][$name])` check should also make sure the config section isn't a plugin? So it would skip if `General` or `database_tests` is used, but not if `ExamplePluginSettings` is used .

**Comment by sgiehl** [comment]
@diosmosis should be fixed now

**Comment by diosmosis** [comment]
👍 merged

**Pull request merged**